docs/DesignDocs/DebuggerSupportForMultiModuleEnclaves.md

```diff
@@ -0,0 +1,268 @@
+  they are called by the SDK's host-side runtime, the breakpoints are hit and
+  the debugger gets control. The debugger then reads the `oe_debug_enclave_t`
+  parameter to handle the notification.
+
+- Windows
+ 
+  Use of breakpoints by the debugger is not recommended. Instead, the Windows
+  `RaiseException` API is used by the runtime to notify the debugger.
+  
+  `oe_debug_notify_enclave_created` raises the following exception:
+  
+  ```c
+        ULONG_PTR args[1] = {(ULONG_PTR)enclave};
+		RaiseException(
+		  OE_DEBUGRT_ENCLAVE_CREATED_EVENT,
+		  0, // dwFlags
+		  1, // number of args
+		  args);
+  ```
+  
+  whereas `oe_debug_notify_enclave_terminated` raises:
+  
+  ```c
+        ULONG_PTR args[1] = {(ULONG_PTR)enclave};
+        RaiseException(
+          OE_DEBUGRT_ENCLAVE_TERMINATED_EVENT,
+          0, // dwFlags
+          1, // number of args
+          args);
+  ```
+  
+  The Windows debuggers (Visual Studio and WinDbg) consume the
+  `OE_DEBUGRT_ENCLAVE_CREATED_EVENT` and `OE_DEBUGRT_ENCLAVE_TERMINATED_EVENT`
+  events. Upon receiving the events, the debuggers invoke the appropriate
+  infrastructure to enable debugging the module.
+
+
+Use-cases for Multi-Module Enclaves
+-----------------------------------
+
+There are a few implementations of multi-module enclaves based on OE SDK:
+
+- [SGX LKL](https://github.com/lsds/sgx-lkl)
+
+  This project runs a modified linux kernel within an enclave and supports running
+  enclaves comprised of executables and shared libraries. This project supports
+  debugging of multi-module enclaves.
+  https://github.com/lsds/sgx-lkl/tree/oe_port/tools/gdb
+
+- [feature.openlibos](https://github.com/openenclave/openenclave/tree/feature.openlibos)
+
+  This branch of OE SDK runs a library OS within an enclave. It also supports
+  enclave comprised of multiple executables and shared libraries.
+  https://github.com/openenclave/openenclave/blob/feature.openlibos/debugger/pythonExtension/gdb_sgx_plugin.py
+
+- [feature/dynamic_binding](https://github.com/openenclave/openenclave/tree/feature/dynamic_binding)
+
+  This branch of OE SDK allows loading and running multi-module enclaves with
+  the goal of permitting use of FIPS certified crypto modules. It also supports
+  debugging multi-module enclaves:
+  https://github.com/openenclave/openenclave/commit/254ae94324b619861f58d0110a1affac37e5a2a4?branch=254ae94324b619861f58d0110a1affac37e5a2a4&diff=split
+
+
+The debugger changes required to support multi-module enclaves are quite similar
+in the above three implementations. The following proposal formally incorporates
+the necessary debugger changes into the Debugger Contract.
+
+Debugger Contract Extensions to Support Multi-Module Enclaves
+-------------------------------------------------------------
+
+The following `oe_debug_module_t` structure captures the information necessary
+to debug a module:
+
+```c
+
+    typedef struct _debug_module_t
```

Comment:
Similar to the _debug_enclave_t structure, there will need to be a list of the modules that have already been loaded stored somewhere. That way a debugger that attaches can still determine the set of modules without having received the create events for them.